lib/Serialization/DeserializeSIL.cpp

```diff
@@ -3280,6 +3296,90 @@ SILDeserializer::lookupDefaultWitnessTable(SILDefaultWitnessTable *existingWt) {
+
+  BCOffsetRAII restoreOffset(SILCursor);
+  SILCursor.JumpToBit(diffWitnessOrOffset.getOffset());
+  auto entry = SILCursor.advance(AF_DontPopBlockAtEnd);
+  if (entry.Kind == llvm::BitstreamEntry::Error) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Cursor advance error in readDifferentiabilityWitness.\n");
+    return nullptr;
+  }
+
+  SmallVector<uint64_t, 64> scratch;
+  StringRef blobData;
+  unsigned kind = SILCursor.readRecord(entry.ID, scratch, &blobData);
+  assert(kind == SIL_DIFFERENTIABILITY_WITNESS &&
+         "Expected sil_differentiability_witness");
+  (void)kind;
+
+  DeclID originalNameId, jvpNameId, vjpNameId;
+  unsigned rawLinkage, isSerialized, numParameterIndices, numResultIndices;
+  GenericSignatureID derivativeGenSigID;
+  ArrayRef<uint64_t> rawParameterAndResultIndices;
+
+  DifferentiabilityWitnessLayout::readRecord(
+      scratch, originalNameId, rawLinkage, isSerialized, derivativeGenSigID,
+      jvpNameId, vjpNameId, numParameterIndices, numResultIndices,
+      rawParameterAndResultIndices);
+
+  auto linkage = fromStableSILLinkage(rawLinkage);
+  assert(linkage && "Expected value linkage for sil_differentiability_witness");
+  auto originalName = MF->getIdentifier(originalNameId).str();
+  auto jvpName = MF->getIdentifier(jvpNameId).str();
+  auto vjpName = MF->getIdentifier(vjpNameId).str();
+  auto *original = getFuncForReference(originalName);
+  auto *jvp = getFuncForReference(jvpName);
+  auto *vjp = getFuncForReference(vjpName);
+  auto derivativeGenSig = MF->getGenericSignature(derivativeGenSigID);
+
+  SmallVector<unsigned, 8> parameterAndResultIndices(
+      rawParameterAndResultIndices.begin(),
+      rawParameterAndResultIndices.end());
+  assert(parameterAndResultIndices.size() ==
+             numParameterIndices + numResultIndices &&
+         "Parameter/result indices count mismatch");
+  auto *parameterIndices = IndexSubset::get(
+      MF->getContext(), original->getLoweredFunctionType()->getNumParameters(),
+      ArrayRef<unsigned>(parameterAndResultIndices)
+          .take_front(numParameterIndices));
+  auto *resultIndices = IndexSubset::get(
+      MF->getContext(), original->getLoweredFunctionType()->getNumResults(),
+      ArrayRef<unsigned>(parameterAndResultIndices)
+          .take_back(numResultIndices));
+
+  auto *diffWitness = SILDifferentiabilityWitness::create(
+      SILMod, *linkage, original, parameterIndices, resultIndices,
+      derivativeGenSig, jvp, vjp, isSerialized);
+  diffWitnessOrOffset.set(diffWitness, /*isFullyDeserialized*/ true);
+  return diffWitness;
+}
+
+SILDifferentiabilityWitness *SILDeserializer::lookupDifferentiabilityWitness(
+    StringRef mangledDiffWitnessKey) {
+  if (!DifferentiabilityWitnessList)
+    return nullptr;
+  auto iter = DifferentiabilityWitnessList->find(mangledDiffWitnessKey);
+  if (iter == DifferentiabilityWitnessList->end())
+    return nullptr;
+
+  auto *diffWitness = readDifferentiabilityWitness(*iter);
```

Comment:
Return this directly?

Reply:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.